File: neet_code_150_cpp/trees/subtree_of_another_tree.cpp
#include <iostream>

struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;

    TreeNode()
        :val(0), left(nullptr), right(nullptr) {}

    TreeNode(int x)
        :val(x), left(nullptr), right(nullptr) {}

    TreeNode(int x, TreeNode* left, TreeNode* right)
        :val(x), left(left), right(right) {}
};

void show_tree_depth(TreeNode* root) {
    if(!root) {
        return;
    }

    std::cout << root->val << std::endl;;

    // left
    show_tree_depth(root->left);
    // right
    show_tree_depth(root->right);
}

// 解けなかった
class Solution {
    public:
        bool isSubtree(TreeNode* root, TreeNode* subRoot) {
            if(!root && !subRoot) {
                return true;
            } else if(!root || !subRoot) {
                return false;
            }

            if(root->val == subRoot->val) {
                return isSamaTree(root, subRoot);
            }

            return isSubtree(root->left, subRoot) && isSubtree(root->right, subRoot);
        }

    private:
        bool isSamaTree(TreeNode* p, TreeNode* q) {
            if(!p && !q) {
                return true;
            } else if(!p || !q) {
                return false;
            }

            if(p->val != q->val) {
                return false;
            }

            return isSamaTree(p->left, q->left) && isSamaTree(p->right, q->right);
        }
};

// 模範解答
class SolutionAns {
    public:
        bool isSubtree(TreeNode* root, TreeNode* subRoot) {
            if (root == NULL) {
                return false;
            }
            if (isSame(root, subRoot)) {
                return true;
            }
            return isSubtree(root->left, subRoot) || isSubtree(root->right, subRoot);
        }

    private:
        bool isSame(TreeNode* root, TreeNode* subRoot) {
            if (root == NULL && subRoot == NULL) {
                return true;
            }
            if (root == NULL || subRoot == NULL) {
                return false;
            }
            if (root->val != subRoot->val) {
                return false;
            }
            return isSame(root->left, subRoot->left) && isSame(root->right, subRoot->right);
        }
};

int main(void) {
    TreeNode root_1(3);
    TreeNode root_1_1(4);
    TreeNode root_1_2(5);
    TreeNode root_1_3(1);
    TreeNode root_1_4(2);
    root_1.left = &root_1_1;
    root_1.right = &root_1_2;
    root_1_1.left = &root_1_3;
    root_1_1.right = &root_1_4;
    TreeNode subRoot_1(4);
    TreeNode subRoot_1_1(1);
    TreeNode subRoot_1_2(2);
    subRoot_1.left = &subRoot_1_1;
    subRoot_1.right = &subRoot_1_2;
   /* 
           root      subRoot 
          
            3         4
           / \       / \
          /   \     /   \
         4     5   1     2 
        / \
       /   \
      1     2 

     =>  true
   */

    TreeNode root_2(3);
    TreeNode root_2_1(4);
    TreeNode root_2_2(5);
    TreeNode root_2_3(1);
    TreeNode root_2_4(2);
    root_1.left = &root_1_1;
    root_1.right = &root_1_2;
    root_1_1.left = &root_1_3;
    root_1_1.right = &root_1_4;
    TreeNode subRoot_2(4);
    TreeNode subRoot_2_1(1);
    TreeNode subRoot_2_2(2);
    subRoot_1.left = &subRoot_1_1;
    subRoot_1.right = &subRoot_1_2;
   /* 
           root      subRoot 
          
            3         4
           / \       / \
          /   \     /   \
         4     5   1     2 
        / \
       /   \
      1     2 
           /
          /
         0

     => false 
   */

    Solution s_1;

    std::cout << s_1.isSubtree(&root_1, &subRoot_1) << std::endl;
    std::cout << s_1.isSubtree(&root_2, &subRoot_2) << std::endl;
}

File: neet_code_150/src/trees/subtree_of_another_tree.rs
use std::{cell::RefCell, rc::Rc};

// Definition for a binary tree node.
#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

fn gen_tree_dfs_2(i: usize, array: &[Option<i32>], root: &Option<Rc<RefCell<TreeNode>>>) {
    if let Some(node) = root {
        if i == 0 {
            node.borrow_mut().val = array[0].unwrap();
        }

        if array.len() > 2 * i + 1 {
            node.borrow_mut().left = if let Some(value) = array[2 * i + 1] {
                Some(Rc::new(RefCell::new(TreeNode::new(value))))
            } else {
                None
            };
            gen_tree_dfs_2(2 * i + 1, array, &node.borrow_mut().left);
        }

        if array.len() > 2 * i + 2 {
            node.borrow_mut().right = if let Some(value) = array[2 * i + 2] {
                Some(Rc::new(RefCell::new(TreeNode::new(value))))
            } else {
                None
            };
            gen_tree_dfs_2(2 * i + 2, array, &node.borrow_mut().right);
        }
    }
}

struct Solution {}
impl Solution {
    // まだACしていない途中
    pub fn is_subtree(
        root: Option<Rc<RefCell<TreeNode>>>,
        sub_root: Option<Rc<RefCell<TreeNode>>>,
    ) -> bool {
        match (root, sub_root) {
            (None, _) => false,

            (Some(node_root), None) => {
                let mut node_root_ref = node_root.borrow_mut();
                Self::is_subtree(node_root_ref.left.take(), None)
                    || Self::is_subtree(node_root_ref.right.take(), None)
            }

            (Some(node_root), Some(node_sub_root)) => {
                if Self::is_same(Some(Rc::clone(&node_root)), Some(Rc::clone(&node_sub_root))) {
                    return true;
                }

                let mut node_root_ref = node_root.borrow_mut();
                Self::is_subtree(node_root_ref.left.take(), Some(Rc::clone(&node_sub_root)))
                    || Self::is_subtree(node_root_ref.right.take(), Some(Rc::clone(&node_sub_root)))
            }
        }
    }

    fn is_same(
        root: Option<Rc<RefCell<TreeNode>>>,
        subroot: Option<Rc<RefCell<TreeNode>>>,
    ) -> bool {
        match (root, subroot) {
            (None, None) => true,
            (Some(_node_p), None) => false,
            (None, Some(_node_q)) => false,

            (Some(node_p), Some(node_q)) => {
                let mut node_p_ref = node_p.borrow_mut();
                let mut node_q_ref = node_q.borrow_mut();
                if node_p_ref.val != node_q_ref.val {
                    return false;
                }

                Self::is_same(node_p_ref.left.take(), node_q_ref.left.take())
                    && Self::is_same(node_p_ref.right.take(), node_q_ref.right.take())
            }
        }
    }
}

fn main() {
    let root_1_1 = Some(Rc::new(RefCell::new(TreeNode::new(0))));
    gen_tree_dfs_2(0, &[Some(3), Some(4), Some(5), Some(1), Some(2)], &root_1_1);
    let root_1_2 = Some(Rc::new(RefCell::new(TreeNode::new(0))));
    gen_tree_dfs_2(0, &[Some(4), Some(1), Some(2)], &root_1_2);

    // => true
    println!("root_1: {}", Solution::is_subtree(root_1_1, root_1_2));

    let root_2_1 = Some(Rc::new(RefCell::new(TreeNode::new(0))));
    // [3,4,5,1,2,null,null,null,null,0]
    gen_tree_dfs_2(
        0,
        &[
            Some(3),
            Some(4),
            Some(5),
            Some(1),
            Some(2),
            None,
            None,
            None,
            None,
            Some(0),
        ],
        &root_2_1,
    );
    let root_2_2 = Some(Rc::new(RefCell::new(TreeNode::new(0))));
    gen_tree_dfs_2(0, &[Some(4), Some(1), Some(2)], &root_2_2);

    // => false
    println!("root_2: {}", Solution::is_subtree(root_2_1, root_2_2));

    let root_3_1 = Some(Rc::new(RefCell::new(TreeNode::new(0))));
    gen_tree_dfs_2(0, &[Some(1), Some(1)], &root_3_1);
    let root_3_2 = Some(Rc::new(RefCell::new(TreeNode::new(0))));
    gen_tree_dfs_2(0, &[Some(1)], &root_3_2);

    // => true
    // このケースがACしない
    println!("root_3: {}", Solution::is_subtree(root_3_1, root_3_2));
}
